src/main/java/com/vmware/vipclient/i18n/messages/api/opt/local/LocalLocaleOpt.java

```diff
@@ -42,27 +57,98 @@
 	    		}
 			} else {
 				path = Paths.get(uri);
-				getSupportedLocales(path, supportedLocales, displayLanguage);
+				getSupportedLocales(path, supportedLocales, languagesNames);
 			}
 	    	
 		} catch (Exception e) {
 			logger.debug(e.getMessage());
 		}
     	return supportedLocales;
     }
-    
-    private void getSupportedLocales(Path path, Map<String, String> supportedLocales, String displayLanguage) throws IOException {
-    	Locale inLocale = Locale.forLanguageTag(displayLanguage); 
+
+    private void getSupportedLocales(Path path, Map<String, String> supportedLocales, JSONObject languagesNames) throws IOException {
     	try (Stream<Path> listOfFiles = Files.walk(path).filter(p -> Files.isRegularFile(p))) {
         	listOfFiles.map(file -> {
 				String fileName = file.getFileName().toString();
-				return fileName.substring(BUNDLE_PREFIX.length(), fileName.indexOf('.'));
-			}).forEach(s->supportedLocales.put(s, Locale.forLanguageTag(s).getDisplayName(inLocale)));
+				if(fileName.startsWith(BUNDLE_PREFIX) && fileName.endsWith(BUNDLE_SUFFIX)) {
+					return fileName.substring(BUNDLE_PREFIX.length(), fileName.indexOf('.'));
+				}
+				return "";
+			}).forEach(language -> {
+				if(language != null && !"".equalsIgnoreCase(language) && !ConstantsKeys.SOURCE.equalsIgnoreCase(language)) {
+					if (languagesNames != null) {
+						supportedLocales.put(language, (String) languagesNames.get(language));
+					} else {
+						supportedLocales.put(language, "");
+					}
+				}
+			});
         }
     }
 
+    private JSONObject getLanguagesNames(String displayLanguage){
+		JSONObject languagesNames = null;
+		logger.debug("Look for languages from cache for locale [{}]!", displayLanguage);
+		String cacheKey = LANGUAGES_PREFIX + displayLanguage;
+		languagesNames = new PatternCacheService().lookForPatternsFromCache(cacheKey);// key
+		if (languagesNames != null) {
+			logger.debug("Find languages from cache for locale [{}]!", displayLanguage);
+			return languagesNames;
+		}
+		languagesNames = getLanguagesNamesFromBundle(displayLanguage);
+		if (languagesNames != null) {
+			logger.debug("Find the languages from local bundle for locale [{}].\n", displayLanguage);// [datetime] and
+			new PatternCacheService().addPatterns(cacheKey, languagesNames);
+			logger.debug("Languages is cached for locale [{}]!\n\n", displayLanguage);
+			return languagesNames;
+		}
+		if (!LocaleUtility.isDefaultLocale(displayLanguage)) {
+			logger.info("Can't find languages for locale [{}], look for English languages as fallback!", displayLanguage);
+			languagesNames = getLanguagesNames(LocaleUtility.getDefaultLocale().toLanguageTag());
+		}
+		return languagesNames;
+
+	}
+
+	private JSONObject getLanguagesNamesFromBundle(String displayLanguage) {
+		logger.debug("Look for languages from local package bundle for locale [{}]!", displayLanguage);
+		String normalizedLocale = CommonUtil.getCLDRLocale(displayLanguage, localePathMap, localeAliasesMap);
```

Comment:
I think it is okay to use CommonUtil.getCLDRLocale in the data (opt) layer. Also, what does it have to do with "pattern".

Reply:
This is locale related code, so your suggestion should be 'do this in LocaleMessage or LocaleService.', right?

The reason of can't resolve it now is the same with below comment:

Don't need this method because client should get the locale from language and region in offline mode.
In general, Getting Pattern by locale is enough.


Currently locale fallback/normalization is based on cldr locale related data. So this means L2 data needed for locale fallback/normalization must be always included in Java client package.

But currently to reduce the size of Java client package, pattern data is put into a separate jar, and for online mode this jar isn't neccessary. If product choose online mode and don't add this jar, change as your suggestion will lead to program broken.

So I prefer to remove this after we can guarantee data needed for locale fallback/normalization is always included or we define new locale fallback/normalization rule which doesn't relies on cldr data.